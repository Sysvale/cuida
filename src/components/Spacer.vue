<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<div class="spacer">
		<slot />
	</div>
</template>

<script>
export default {
	props: {
		padding: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		paddingTop: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		paddingRight: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		paddingBottom: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		paddingLeft: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		margin: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		marginTop: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		marginRight: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		marginBottom: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
		marginLeft: {
			type: [Number, String],
			default: 0,
			validator: (value) => (value >= 0 && value <= 12),
		},
	},

	computed: {
		paddingTopResolver() {
			let pt = this.paddingTop > 12 ? 12 : +this.paddingTop;
			pt = this.padding ? this.basePropertyResolver(this.padding) : pt;

			return `${pt * 4}px`;
		},

		paddingRightResolver() {
			let pr = this.paddingRight > 12 ? 12 : +this.paddingRight;
			pr = this.padding ? this.basePropertyResolver(this.padding) : pr;
			return `${pr * 4}px`;
		},

		paddingBottomResolver() {
			let pb = this.paddingBottom > 12 ? 12 : +this.paddingBottom;
			pb = this.padding ? this.basePropertyResolver(this.padding) : pb;
			return `${pb * 4}px`;
		},

		paddingLeftResolver() {
			let pl = this.paddingLeft > 12 ? 12 : +this.paddingLeft;
			pl = this.padding ? this.basePropertyResolver(this.padding) : pl;
			return `${pl * 4}px`;
		},

		marginTopResolver() {
			let mt = this.marginTop > 12 ? 12 : +this.marginTop;
			mt = this.margin ? this.basePropertyResolver(this.margin) : mt;
			return `${mt * 4}px`;
		},

		marginRightResolver() {
			let mr = this.marginRight > 12 ? 12 : +this.marginRight;
			mr = this.margin ? this.basePropertyResolver(this.margin) : mr;
			return `${mr * 4}px`;
		},

		marginBottomResolver() {
			let mb = this.marginBottom > 12 ? 12 : +this.marginBottom;
			mb = this.margin ? this.basePropertyResolver(this.margin) : mb;
			return `${mb * 4}px`;
		},

		marginLeftResolver() {
			let ml = this.marginLeft > 12 ? 12 : +this.marginLeft;
			ml = this.margin ? this.basePropertyResolver(this.margin) : ml;
			return `${ml * 4}px`;
		},
	},

	methods: {
		basePropertyResolver(baseProperty) {
			if (baseProperty >= 0 && baseProperty <= 12) {
				return baseProperty;
			}

			if (baseProperty < 0) {
				return 0;
			}

			return 12;
		}
	},
}
</script>

<style lang="scss" scoped>
.spacer {
	width: 100%;
	margin-bottom: v-bind(marginBottomResolver);
	margin-left: v-bind(marginLeftResolver);
	margin-right: v-bind(marginRightResolver);
	margin-top: v-bind(marginTopResolver);

	padding-bottom: v-bind(paddingBottomResolver);
	padding-left: v-bind(paddingLeftResolver);
	padding-right: v-bind(paddingRightResolver);
	padding-top: v-bind(paddingTopResolver);
}
</style>