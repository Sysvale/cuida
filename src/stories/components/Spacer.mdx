import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import Spacer from '../../components/Spacer.vue';
import * as SpacerStories from './Spacer.stories';

<Meta of={SpacerStories} />

export const Template = (args) => ({
	components: { CdsSpacer: Spacer },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<div style="width: fit-content; background-color: #def7eb; border-radius: 20px">
			<div
				style="padding: 0.5px; background-color: #def7eb; border-radius: 20px"
			>
				<cds-spacer
					v-bind="args"
				>
					<div style="width: 100px; height: 100px; background-color: #2AC092; border-radius: 20px"/>
				</cds-spacer>
			</div>
		</div>
	`,
	methods: {
		logClickableClick(event) {
			console.info('logClick: ', event);
		}
	},
});

# Spacer

### Componente auxiliar utilizado para adicionar paddings e margins seguindo o padrão de tokens do Cuida.

***

<br />

## Quando usar:

* No caso de determinado componente ou elemento da interface precisar apenas de espaçamento (padding ou margin) como modificar de estilo.

<br />

## Quando não usar:

* Nos casos em que determinado componente precisar de vários modificadores de estilo. Pra esses casos
  utilize classes e nelas inclua paddings e margins com os tokens de espaçamento do Cuida.

<br />

## Observações:

* ⚠️ Dê preferência sempre que possível à utilização dos tokens em detrimento ao Spacer.
* O valor mandado para as props do Spacer é sempre multiplicado por 4, assim como nos tokens scss do Cuida.
* O valor máximo aceito é 12.
* O spacer não implementa margins e paddings negativos.

## Preview

<Canvas>
  <Story of={SpacerStories.Spacer} />
</Canvas>

<Controls />
