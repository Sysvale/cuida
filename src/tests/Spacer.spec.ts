// @vitest-environment jsdom
import { describe, test, expect } from 'vitest';
import Spacer from '../components/Spacer.vue';
import { mount } from '@vue/test-utils';

describe('Spacer snapshot test', () => {
	test('renders correctly', () => {
		const wrapper = mount(Spacer, {
			props: {
				paddingTop: 0,
				paddingRight: 0,
				paddingBottom: 0,
				paddingLeft: 0,
				marginTop: 0,
				marginRight: 0,
				marginBottom: 0,
				marginLeft: 0,
			},
			slots: {
				default: '<div style="width: 100px; height: 100px; background-color: #2AC092; border-radius: 20px"/>'
			}
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
