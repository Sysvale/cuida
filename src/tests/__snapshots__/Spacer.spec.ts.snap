// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spacer snapshot test > renders correctly 1`] = `
"<div data-v-9cc3ee57="" class="spacer">
  <div style="width: 100px; height: 100px; background-color: rgb(42, 192, 146); border-radius: 20px;"></div>
</div>"
`;
